net/dns: teach OpenBSD's manager to talk to resolvd(8).

qbit opened the pull request:
OpenBSD 6.9 and up has a daemon which handles nameserver configuration. This MR
teaches the OpenBSD dns manager to check if resolvd is being used. If it is, it
will use the route(8) command to tell resolvd to add the Tailscale dns entries
to resolv.conf

Signed-off-by: Aaron Bieber <aaron@bolddaemon.com>

bradfitz commented:
Thanks! Assigning to @danderson for review.

qbit commented:
After putting this together I realized that OpenBSD 6.9's `route` doesn't have the `nameserver` command.  So this will break 6.9 machines. I'll have to think about how to keep compat with 6.9.

danderson commented on net/dns/manager_openbsd.go:
```diff
@@ -4,8 +4,72 @@
+		})
+}
+
+// newOSConfigEnv
```
// newOSConfigEnv are the funcs newOSConfigurator needs, pulled out for testing.

danderson commented on net/dns/manager_openbsd.go:
```diff
@@ -4,8 +4,72 @@
+	}
+}
+
+func rcIsResolvd(resolvConfContents []byte) string {
```
I'd make this func return a bool if it's making a binary decision. I believe you patterned this on linux's resolvOwner, which has to return an N-way decision between all the different managers. OpenBSD is mercifully simpler :)

danderson commented on net/dns/resolvd.go:
```diff
@@ -0,0 +1,53 @@
+	return cmd.Run()
+}
+func (m *resolvdManager) SupportsSplitDNS() bool {
+	return true
```
I don't think this is correct, based on the implementation of SetDNS. Split DNS support here means that the overall manager can tell the OS to only use a set of resolvers for specific domain suffixes. The implementation of SetDNS doesn't make use of MatchDomains, so it wouldn't implement the configuration that the DNS manager wanted.

Can resolvd do per-suffix DNS routes? If so, this can be true, but SetDNS needs to obey MatchDomains in what it configures.

qbit replied:
I don't think resolvd can do per-suffix DNS routes. I am not completely sure I understand DNSR, but it seems that resolved's approach of ordering nameservers based off route priority _almost_ achieves the same thing (but maybe that's just my lack of understanding :P).

I guess making this `false` is the correct path forward.

danderson commented on net/dns/resolvd.go:
```diff
@@ -0,0 +1,53 @@
+		args = append(args, s.String())
+	}
+
+	cmd := exec.Command("/sbin/route", args...)
```
Is this constructed command safe to invoke repeatedly, with either the same nameservers or a different set? I assume so, but want to explicitly confirm since I don't have an OpenBSD box to test on.

qbit replied:
Yep, it adds/removes accordingly.

```
qbit@tal[0]:~$ cat /etc/resolv.conf                                                                                                                                                                         
nameserver 100.100.100.100 # resolvd: tun0
nameserver 10.6.0.2 # resolvd: em0
search qbit.github.beta.tailscale.net
qbit@tal[0]:~$ doas route nameserver tun0 100.100.100.100 
qbit@tal[0]:~$ cat /etc/resolv.conf                       
nameserver 100.100.100.100 # resolvd: tun0
nameserver 10.6.0.2 # resolvd: em0
search qbit.github.beta.tailscale.net
qbit@tal[0]:~$ doas route nameserver tun0 100.100.100.100 1.1.1.1 
qbit@tal[0]:~$ cat /etc/resolv.conf                               
nameserver 1.1.1.1 # resolvd: tun0
nameserver 100.100.100.100 # resolvd: tun0
nameserver 10.6.0.2 # resolvd: em0
search qbit.github.beta.tailscale.net
qbit@tal[0]:~$ doas route nameserver tun0 100.100.100.100         
qbit@tal[0]:~$ cat /etc/resolv.conf                       
nameserver 100.100.100.100 # resolvd: tun0
nameserver 10.6.0.2 # resolvd: em0
search qbit.github.beta.tailscale.net
qbit@tal[0]:~$ 
```

danderson commented on net/dns/resolvd.go:
```diff
@@ -0,0 +1,53 @@
+	ifName string
+}
+
+func (m *resolvdManager) SetDNS(config OSConfig) error {
```
SetDNS also needs to propagate config.SearchDomains into the OS, for single-label name expansion to work right.

qbit replied:
Ok, updated to propagate SearchDomains.

danderson commented on net/dns/resolvd.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+func (m *resolvdManager) GetBaseConfig() (OSConfig, error) {
+	return OSConfig{}, ErrGetBaseConfigNotSupported
```
If resolvd doesn't support per-suffix DNS routing, this needs to be able to read the base OS configuration, so that we can emulate split DNS within tailscale. Not necessary if SupportsSplitDNS() is true and SetDNS supports it natively though.

qbit replied:
GetBaseConfig now returns a partially populated OSConfig. I only had it grab the SearchDomains since resolvd handles the rest.

danderson commented on net/dns/resolvd.go:
```diff
@@ -0,0 +1,53 @@
+	return OSConfig{}, ErrGetBaseConfigNotSupported
+}
+
+func (m *resolvdManager) Close() error {
```
Do you need to clear any settings here? Or does resolvd clean up when the interface vanishes?

qbit replied:
`resolvd` handles cleanup when the interface is removed.

danderson reviewed (commented):
Thanks for the PR! Main feedback is that this OSConfigurator doesn't implement our definition of split DNS, so either it needs to do that (if resolvd supports it), or this has to be able to read + entirely replace the prior OS configuration so that we can emulate split DNS within tailscaled.

The manpages for resolvd and unwind are a bit light, but afaict resolvd supports neither overriding all other DNS configuration, nor per-suffix DNS routing. As such, I'm worried that resolvd cannot implement either of the two behaviors we require to make DNS work correctly :/ I hope I'm wrong and the manpages are just missing the bits I'm looking for.

qbit commented:
OK, I think I resolved everything (possibly introduced some new things too :P)!

qbit commented:
I believe I have addressed all the issues for this PR. I have a squashed version ready to push but I wanted to leave the history for anyone wanting to review again.

Should I just push the squash'd version?

bradfitz commented:
We're all back from vacation now. Leaving for @danderson to review again. I pushed the button to run the CI.

qbit commented:
Awesome, hope it was a good vacation! Thanks for pushin the button!
